### ISSUE TITLE
Lab start crash in stable and experimental

### ISSUE THREAD

**Issue by birddollar** [CleverRaven/Cataclysm-DDA#45469]
# Describe the bug
After a few ingame minutes, the attached save file will experience a crash.

# Steps To Reproduce

1. Several hours into lab start.
2. Reach point in attached save.
3. Pass time in some way, ie read, wait several minutes, or walk in circles.
4. Game will crash after error message.

I also attempted to switch save among CDDA versions, trying out O.E, OE1, OE2, and currently experimental. All of them experienced crashes, seemingly with minor variance on when crash occurred. 

# Expected behavior

Game does not crash.

# Versions and configuration

Windows 10
0.E-7587-g8d87f15 (Also Stable O.E, OE1, OE2)
Tiles, Undead People tileset
English
Mods loaded:: dda, no npc food, Aftershock, bionic professions, alt map key

# Additional context

Crash initially occurred on stable O.E2, currently build is for experimental after someone on reddit suggested I switch it to experimental for the Github bug report. Thus, loading into the save initially results in a bunch of error messages related to the feature differences between builds. After bypassing them and loading into the save, conditions are pretty much the same for all versions to reproduce the crash.

[crash.log](https://github.com/CleverRaven/Cataclysm-DDA/files/5554492/crash.log)
[debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/5554493/debug.log)

[Crashsavebirddollar.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5554515/Crashsavebirddollar.zip)

**Comment by anothersimulacrum** [comment]
Some sort of bug with fd_push_items.
```
#0  0x00007ffff7917615 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff7900862 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff7900747 in __assert_fail_base.cold () from /usr/lib/libc.so.6
#3  0x00007ffff790fbf6 in __assert_fail () from /usr/lib/libc.so.6
#4  0x0000555555e45cde in cata::colony<item, std::allocator<item>, unsigned short>::colony_iterator<false>::operator++ (this=0x0) at src/colony.h:280
#5  0x00005555565db663 in cata::colony<item, std::allocator<item>, unsigned short>::colony_iterator<false>::operator++ (this=0x7fffffffd320) at src/colony.h:300
#6  map::process_fields_in_submap (this=0x55555771fa50, current_submap=0x5555690f6ed0, submap=...)
    at src/map_field.cpp:685
#7  0x00005555565da0fd in map::process_fields (this=0x55555771fa50) at src/map_field.cpp:156
#8  0x00005555562240cc in game::do_turn (this=<optimized out>) at src/game.cpp:1574
#9  0x000055555657aea4 in main (argc=<optimized out>, argv=<optimized out>) at src/main.cpp:707
(gdb) f 6
#6  map::process_fields_in_submap (this=0x55555771fa50, current_submap=0x5555690f6ed0, submap=...)
    at src/map_field.cpp:685
685	                            pushee++;
(gdb) l
680	                if( cur_fd_type_id == fd_push_items ) {
681	                    map_stack items = i_at( p );
682	                    for( auto pushee = items.begin(); pushee != items.end(); ) {
683	                        if( pushee->typeId() != itype_rock ||
684	                            pushee->age() < 1_turns ) {
685	                            pushee++;
686	                        } else {
687	                            item tmp = *pushee;
688	                            tmp.set_age( 0_turns );
689	                            pushee = items.erase( pushee );
```

**Comment by andrei8l** [comment]
If I understand it right, `add_item_or_charges( newp, tmp )` invalidates the iterator `pushee` if this random point `newp` happens to be the same as `p` (i.e. it points to the same backend colony)

```diff
diff --git a/src/map_field.cpp b/src/map_field.cpp
index 70f9350c1b..810e27e9e7 100644
--- a/src/map_field.cpp
+++ b/src/map_field.cpp
@@ -695,6 +695,9 @@ void map::process_fields_in_submap( submap *const current_submap,
                             }
                             if( !valid.empty() ) {
                                 tripoint newp = random_entry( valid );
+                                if( newp == p ) {
+                                    continue;
+                                }
                                 add_item_or_charges( newp, tmp );
                                 if( player_character.pos() == newp ) {
                                     add_msg( m_bad, _( "A %s hits you!" ), tmp.tname() );
```
seems to fix the crash for me - I waited 6 hours from the save file 3 times without a crash.

**Comment by birddollar** [comment]
Thanks so much for the prompt response and solution to this post! Any chance you could ELI5 what I have to do to bypass the crash? I don't have the coding literacy to understand your solution.

**Comment by andrei8l** [comment]
@birddollar you'll have to wait for the patch to get reviewed, unless you're willing to compile it yourself (ELI10-ish instructions for windows are [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/COMPILING/COMPILING-VS-VCPKG.md) - you need to pull from my PR though).
